Source/WebKit/UIProcess/API/glib/WebKitUserContent.cpp

```diff
@@ -433,3 +437,122 @@ API::ContentRuleList& webkitUserContentFilterGetContentRuleList(WebKitUserConten
     ASSERT(userContentFilter);
     return *userContentFilter->contentRuleList;
 }
+
+/**
+ * WebKitUserScriptMessage: (ref-func webkit_user_script_message_ref) (unref-func webkit_user_script_message_unref)
+ *
+ * A helper class that sends user provided script to the signal handler.
+ *
+ * Since: 2.38
+ */
+struct _WebKitUserScriptMessage {
+    _WebKitUserScriptMessage(WebCore::SerializedScriptValue& serializedScriptValue, WTF::Function<void(API::SerializedScriptValue*, const String&)>&& completionHandler)
+        : completionHandler(WTFMove(completionHandler))
+        , referenceCount(1)
+    {
+        JSCContext* jscContext = SharedJavascriptContext::singleton().getOrCreateContext();
+        JSValueRef jsValueRef = serializedScriptValue.deserialize(jscContextGetJSContext(jscContext), nullptr);
+        message = jscValueCreate(jscContext, jsValueRef);
+    }
```

Comment:
We are missing a destructor here: if `completionHandler` has never been used, that means `webkit_user_script_message_reply()` or `webkit_user_script_message_error()` have _not_ been called. In that case, the destructor needs to construct an undefined JS value and send it as reply.

To know whether `completionHandler` has been used or not, we can reset it to `nullptr` after using it. Then, if `completionHandler` is non-null, that means it has not been used.

```suggestion
    }
    
    ~_WebKitUserScriptMessage()
    {
        if (completionHandler) {
            // Make a JSCValue for "undefined" and send it as the reply.
        }
    }
```